go/viperutil/internal/sync/sync.go

```diff
@@ -0,0 +1,302 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"vitess.io/vitess/go/viperutil/internal/log"
+)
+
+// Viper is a wrapper around a pair of viper.Viper instances to provide config-
+// reloading in a threadsafe manner.
+//
+// It maintains one viper, called "disk", which does the actual config watch and
+// reload (via viper's WatchConfig), and a second viper, called "live", which
+// Values (registered via viperutil.Configure with Dynamic=true) access their
+// settings from. The "live" config only updates after blocking all values from
+// reading in order to swap in the most recently-loaded config from the "disk".
+type Viper struct {
+	m    sync.Mutex // prevents races between loadFromDisk and AllSettings
+	disk *viper.Viper
+	live *viper.Viper
+	keys map[string]*sync.RWMutex
+
+	subscribers    []chan<- struct{}
+	watchingConfig bool
```

Comment:
I think we might want to use an atomic.Bool here as at least in theory you could have concurrent Watch or AdaptGetter calls where one is setting it as the other is reading it. No?

Reply:
nope, we check before watching, and that is supposed to be called from `servenv` during parsing. if you do that multiple times, i'd consider that undefined behavior (like if you tried to call `servenv.Run` twice)